util/src/main/java/tech/pegasys/artemis/util/backing/view/AbstractBasicView.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.util.backing.view;
+
+import tech.pegasys.artemis.util.backing.BasicView;
+import tech.pegasys.artemis.util.backing.ViewWrite;
+import tech.pegasys.artemis.util.backing.tree.TreeNode;
+import tech.pegasys.artemis.util.backing.type.BasicViewType;
+
+public abstract class AbstractBasicView<C, V extends AbstractBasicView<C, V>>
+    implements BasicView<C> {
+  private final BasicViewType<V> type;
+  private final C value;
+
+  public AbstractBasicView(C value, BasicViewType<V> type) {
+    this.type = type;
+    this.value = value;
+  }
+
+  @Override
+  public C get() {
+    return value;
+  }
+
+  @Override
+  public BasicViewType<V> getType() {
+    return type;
+  }
+
+  @Override
+  public TreeNode getBackingNode() {
+    return getType().createTreeNode(getThis());
+  }
+
+  @SuppressWarnings("unchecked")
+  protected V getThis() {
+    return (V) this;
+  }
+
+  @Override
+  public ViewWrite createWritableCopy() {
+    throw new UnsupportedOperationException("Basic view instances are immutable");
```

Comment:
Is it possible to break up the interfaces in a more granular way so we don't end up needing these `UnsupportedOperationException`'s ?

Reply:
Very good question! :)
I honestly tried to do this, but didn't succeeded. The reason is that mutable 'views' extends immutable, like this: 
```java
interface ImmutableA {
    int getInt();
}
interface MutableA extends ImmutableA {
    void setInt(int i);
}

interface ImmutableB {
    ImmutableA getA();
}
interface MutableB {
    // here we can narrow return type to mutable
    MutableA getA();
}
```
This way you get the whole data hierarchy either immutable or mutable. 
Separating immutable and mutable class hierarchies is possible though eliminates flexibility. 
E.g. beacon state transition functions requires `MutableBeaconState` but they use functions which need just readonly `BeaconState`. In the current design you just pass `MutableBeaconState` instance (which it subclass of `BeaconState`). If hierarchies are separated it would become less graceful. 

Oh, just realized that I was explaining similar issue which is a bit different from your concern: my wish to prohibit the following methods for `ViewWrite`:
- `createWriteableCopy`
- `hashTreeRoot`

All my efforts to achieve finer interface granularity and generic type safety had resulted in a pretty complex and large interfaces set, so I just gave up finally 

Any ideas on improvements which don't lead to number of interfaces explosure are very welcome!